### PR TITLE
Feature/simd overflow checks

### PR DESCRIPTION
**Pull request by markos**
Implemented the so-called 'Safe Math' feature for Fixed32/Fixed64, both scalar and AVX2 code paths.
In particular, on Overflow/Underflow, set the respective flag in the FixedPoint fp_state and return FP_MAX/FP_MIN. In case of NaN/+inf/-inf, also set the resp. flags and return the appropriate value. All basic vectorised operations have been modified to behave the same. 
TODO: Vectorise the math functions (exp/log/trig/etc) for performance but also keep the same behaviour as scalar.
Increased test coverage for both scales.
TODO: It was not possible to provide a clean way to select both Safe and Unsafe FixedPoint at the same time -using a run-time switch or providing two different data types (eg. fp32_unsafe_t). This will be dealt in a different PR.

### CODE REVIEW

File: libs/vectorise/tests/gtest/avx_tests.cpp [hunk 1]
```diff
@@ -99,10 +100,12 @@ using MyTypes = ::testing::Types<
     fetch::vectorise::VectorRegister<double, 128>, fetch::vectorise::VectorRegister<double, 256>>;
 
 using MyFPTypes =
-    ::testing::Types<fetch::vectorise::VectorRegister<float, 256>,
+    ::testing::Types</*fetch::vectorise::VectorRegister<float, 256>,*/
```
Comment (baykaner) [inline libs/vectorise/tests/gtest/avx_tests.cpp]: these types became unsupported?

Comment (markos) [reply libs/vectorise/tests/gtest/avx_tests.cpp]: no, I removed them completely, the particular test is for checking the NaN/Inf/etc states work with SIMD FixedPoint, we already know that these work with floats/doubles already.

File: libs/math/tests/unit/fixed_point/fixed_point.cpp [hunk 2]
```diff
@@ -85,16 +85,16 @@ TEST(FixedPointTest, Conversion_16_16)
   // Largest fractional closest to one, representable to a FixedPoint
   FixedPoint<16, 16> almost_one(0.99999);
   // Largest fractional closest to one, representable to a FixedPoint
-  FixedPoint<16, 16> largest_int(std::numeric_limits<int16_t>::max());
+  FixedPoint<16, 16> largest_int(std::numeric_limits<int16_t>::max() - 1);
 
   // Smallest possible integer, increase by one, in order to allow for the fractional part.
-  FixedPoint<16, 16> smallest_int(std::numeric_limits<int16_t>::min());
+  FixedPoint<16, 16> smallest_int(std::numeric_limits<int16_t>::min() + 2);
```
Comment (evsmithx) [inline libs/math/tests/unit/fixed_point/fixed_point.cpp:92]: I think the comment above this line needs updating - it's now increased by 2?

Comment (evsmithx) [reply libs/math/tests/unit/fixed_point/fixed_point.cpp:92]: See also the comment on line 87

File: libs/math/tests/unit/fixed_point/fixed_point.cpp [hunk 3]
```diff
@@ -103,13 +103,13 @@ TEST(FixedPointTest, Conversion_16_16)
...
-  EXPECT_EQ(fp32_t::MIN, 0x8000ffff);
-  EXPECT_EQ(fp32_t::MAX, 0x7fffffff);
+  EXPECT_EQ(fp32_t::MIN, 0x80010001);
+  EXPECT_EQ(fp32_t::MAX, 0x7ffeffff);
 
   // We cannot be smaller than the actual negative integer of the actual type
   EXPECT_TRUE(smallest_fixed_point.Data() > std::numeric_limits<int32_t>::min());
   // On the other hand we expect to be exactly the same as the largest positive integer of int64_t
-  EXPECT_TRUE(largest_fixed_point.Data() == std::numeric_limits<int32_t>::max());
+  EXPECT_TRUE(largest_fixed_point.Data() < std::numeric_limits<int32_t>::max());
```
Comment (evsmithx) [inline libs/math/tests/unit/fixed_point/fixed_point.cpp:113]: Update comment

File: libs/math/tests/unit/fixed_point/fixed_point.cpp [hunk 4]
```diff
@@ -184,30 +184,30 @@ TEST(FixedPointTest, Conversion_32_32)
   // Largest fractional closest to one, representable to a FixedPoint
   FixedPoint<32, 32> almost_one(0.9999999998);
   // Largest fractional closest to one, representable to a FixedPoint
-  FixedPoint<32, 32> largest_int(std::numeric_limits<int32_t>::max());
+  FixedPoint<32, 32> largest_int(std::numeric_limits<int32_t>::max() - 1);
 
   // Smallest possible integer, increase by one, in order to allow for the fractional part.
-  FixedPoint<32, 32> smallest_int(std::numeric_limits<int32_t>::min());
+  FixedPoint<32, 32> smallest_int(std::numeric_limits<int32_t>::min() + 2);
```
Comment (evsmithx) [inline libs/math/tests/unit/fixed_point/fixed_point.cpp:192]: Update comment

File: libs/math/tests/unit/fixed_point/fixed_point.cpp [hunk 5]
```diff
@@ -224,6 +224,104 @@ TEST(FixedPointTest, Conversion_32_32)
...
+//   EXPECT_EQ(static_cast<uint32_t>(one), 1);
+//   EXPECT_EQ(static_cast<long>(one), 1);
+//   EXPECT_EQ(static_cast<uint64_t>(one), 1);
+//   EXPECT_EQ(static_cast<int64_t>(one), 1);
+//   EXPECT_EQ(static_cast<uint64_t>(one), 1);
+
+//   EXPECT_EQ(fp128_t::TOLERANCE.Data(), 0x200);
+//   EXPECT_EQ(fp128_t::DECIMAL_DIGITS, 9);
+// }
+
```
Comment (evsmithx) [inline libs/math/tests/unit/fixed_point/fixed_point.cpp]: Remove commented-out code

File: libs/math/tests/unit/vectors/avx_tests.cpp [hunk 6]
```diff
@@ -282,3 +283,158 @@ TYPED_TEST(VectorReduceTest, reduce_tests)
...
+    vret.Store(D);
+    for (size_t i = 0; i < TypeParam::E_BLOCK_COUNT; i++)
+    {
+      if (!type::IsNaN(C[i]) && !type::IsNaN(D[i]))
+      {
+        EXPECT_EQ(C[i], D[i]);
+      }
+    }
+  }
+  std::cout << "done " << j << " NaN/Inf tests" << std::endl;
```
Comment (evsmithx) [inline libs/math/tests/unit/vectors/avx_tests.cpp]: Debugging code

File: libs/vectorise/include/vectorise/arch/avx2/register_fixed64.hpp [hunk 7]
```diff
@@ -238,21 +302,410 @@ inline std::ostream &operator<<(std::ostream &s, __int128_t const &n)
...
-  prod[0]       = a128[0] * b128[0];
-  prod[1]       = a128[1] * b128[1];
-  __m256i vprod = _mm256_load_si256(reinterpret_cast<__m256i const *>(prod));
-  vprod         = _mm256_bsrli_epi128(vprod, 4);
-  vprod         = _mm256_permute4x64_epi64(vprod, 0x8);
-  return {_mm256_extractf128_si256(vprod, 0)};
+  prod128[0]    = (a128[0] * b128[0]) >> 32;
+  prod128[1]    = (a128[1] * b128[1]) >> 32;
+  __m256i vprod = _mm256_load_si256(reinterpret_cast<__m256i const *>(prod128));
+  // vprod         = _mm256_bsrli_epi128(vprod, 4);
```
Comment (evsmithx) [inline libs/vectorise/include/vectorise/arch/avx2/register_fixed64.hpp]: Left from previous version?

File: libs/vectorise/include/vectorise/fixed_point/fixed_point.hpp [hunk 8]
```diff
@@ -20,6 +20,7 @@
 #include "meta/tags.hpp"
 #include "meta/type_traits.hpp"
 #include "vectorise/platform.hpp"
+//#include "vectorise/uint/int.hpp"
```
Comment (evsmithx) [inline libs/vectorise/include/vectorise/fixed_point/fixed_point.hpp]: old code

File: libs/vectorise/include/vectorise/fixed_point/fixed_point.hpp [hunk 9]
```diff
@@ -49,6 +50,18 @@ struct TypeFromSize
...
+// struct TypeFromSize<256>
+// {
+//   static constexpr bool     is_valid = true;
+//   static constexpr uint16_t size     = 256;
+//   using ValueType                    = vectorise::Int<256>;
+//   using UnsignedType                 = vectorise::Int<256>;
+//   using SignedType                   = vectorise::Int<256>;
+//   // using NextSize                        = TypeFromSize<256>;
+// };
+
```
Comment (evsmithx) [inline libs/vectorise/include/vectorise/fixed_point/fixed_point.hpp]: old code

**Review by evsmithx — CHANGES_REQUESTED** [review]
I can't say I really understand all of this, so it should probably be reviewed by someone else!

File: libs/math/tests/unit/fixed_point/fixed_point.cpp (hunk 5 above)
Comment (markos) [reply libs/math/tests/unit/fixed_point/fixed_point.cpp]: fixed.

File: libs/vectorise/include/vectorise/arch/avx2/register_fixed64.hpp (hunk 7 above)
Comment (markos) [reply libs/vectorise/include/vectorise/arch/avx2/register_fixed64.hpp]: removed.

File: libs/vectorise/include/vectorise/fixed_point/fixed_point.hpp (hunk 9 above)
Comment (markos) [reply libs/vectorise/include/vectorise/fixed_point/fixed_point.hpp]: removed.

File: libs/vectorise/include/vectorise/fixed_point/fixed_point.hpp (hunk 8 above)
Comment (markos) [reply libs/vectorise/include/vectorise/fixed_point/fixed_point.hpp]: removed.

File: libs/math/tests/unit/fixed_point/fixed_point.cpp (hunk 2 above)
Comment (markos) [reply libs/math/tests/unit/fixed_point/fixed_point.cpp:92]: resolved.

File: libs/math/tests/unit/fixed_point/fixed_point.cpp (hunk 3 above)
Comment (markos) [reply libs/math/tests/unit/fixed_point/fixed_point.cpp:113]: resolved.

File: libs/math/tests/unit/fixed_point/fixed_point.cpp (hunk 4 above)
Comment (markos) [reply libs/math/tests/unit/fixed_point/fixed_point.cpp:192]: resolved.

File: libs/math/tests/unit/vectors/avx_tests.cpp (hunk 6 above)
Comment (markos) [reply libs/math/tests/unit/vectors/avx_tests.cpp]: removed.

File: libs/math/tests/unit/fixed_point/fixed_point.cpp [hunk 10]
```diff
@@ -224,6 +224,104 @@ TEST(FixedPointTest, Conversion_32_32)
   EXPECT_EQ(fp64_t::DECIMAL_DIGITS, 9);
 }
 
+// TEST(FixedPointTest, Conversion_64_64)
```
Comment (troels-im) [inline libs/math/tests/unit/fixed_point/fixed_point.cpp]: I'd suggest to delete or fix this test.

**Review by troels-im — APPROVED** [review]
LGTM.

File: libs/math/tests/unit/fixed_point/fixed_point.cpp (hunk 10 above)
Comment (markos) [reply libs/math/tests/unit/fixed_point/fixed_point.cpp]: it's actually removed, was accidentally added in this PR

**Review by evsmithx — APPROVED** [review]

**Pull request merged**